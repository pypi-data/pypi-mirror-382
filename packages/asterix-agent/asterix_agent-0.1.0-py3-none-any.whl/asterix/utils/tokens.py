"""
MemGPT Token Counting Utilities

Provides token counting functionality for memory management and eviction decisions.
Supports tiktoken for accurate counting and fallback methods.
"""

import re
import logging
from typing import Dict, List, Optional, Union, Any
from dataclasses import dataclass

try:
    import tiktoken
    TIKTOKEN_AVAILABLE = True
except ImportError:
    TIKTOKEN_AVAILABLE = False
    logging.warning("tiktoken not available, using approximate token counting")

from .config import get_config

logger = logging.getLogger(__name__)


@dataclass
class TokenCount:
    """Token count result with metadata"""
    tokens: int
    characters: int
    words: int
    method: str  # "tiktoken", "approximate", "character_based"
    model: Optional[str] = None


class TokenCounter:
    """
    Token counting utility supporting multiple counting methods.
    
    Provides accurate token counting for memory management and eviction decisions.
    """
    
    def __init__(self):
        self.config = get_config()
        self._encoders: Dict[str, any] = {}
        self._load_encoders()
    
    def _load_encoders(self):
        """Load tiktoken encoders for different models"""
        if not TIKTOKEN_AVAILABLE:
            logger.warning("tiktoken not available, using fallback counting methods")
            return
        
        try:
            # Common encoders
            self._encoders["cl100k_base"] = tiktoken.get_encoding("cl100k_base")  # GPT-4, text-embedding-3
            self._encoders["p50k_base"] = tiktoken.get_encoding("p50k_base")      # GPT-3.5, text-davinci-003
            self._encoders["r50k_base"] = tiktoken.get_encoding("r50k_base")      # GPT-3, text-davinci-002
            
            logger.info("Loaded tiktoken encoders for accurate token counting")
        except Exception as e:
            logger.error(f"Failed to load tiktoken encoders: {e}")
            self._encoders.clear()
    
    def get_encoder_for_model(self, model: str) -> Optional[any]:
        """
        Get appropriate encoder for a given model.
        
        Args:
            model: Model name
            
        Returns:
            Tiktoken encoder or None if not available
        """
        if not TIKTOKEN_AVAILABLE or not self._encoders:
            return None
        
        # Model to encoder mapping
        model_encoders = {
            # OpenAI models
            "gpt-4": "cl100k_base",
            "gpt-4-turbo": "cl100k_base", 
            "gpt-4-turbo-preview": "cl100k_base",
            "gpt-3.5-turbo": "cl100k_base",
            "text-embedding-3-small": "cl100k_base",
            "text-embedding-3-large": "cl100k_base",
            "text-embedding-ada-002": "cl100k_base",
            
            # Legacy models
            "text-davinci-003": "p50k_base",
            "text-davinci-002": "r50k_base",
            "davinci": "r50k_base",
            "curie": "r50k_base",
            "babbage": "r50k_base",
            "ada": "r50k_base"
        }
        
        # Find encoder for model
        for model_name, encoder_name in model_encoders.items():
            if model_name in model.lower():
                return self._encoders.get(encoder_name)
        
        # Default to cl100k_base for unknown models
        return self._encoders.get("cl100k_base")
    
    def count_tokens_tiktoken(self, text: str, model: Optional[str] = None) -> TokenCount:
        """
        Count tokens using tiktoken for maximum accuracy.
        
        Args:
            text: Text to count tokens for
            model: Model name for encoder selection
            
        Returns:
            Token count result
        """
        if not TIKTOKEN_AVAILABLE or not self._encoders:
            return self.count_tokens_approximate(text)
        
        try:
            # Get encoder
            if model:
                encoder = self.get_encoder_for_model(model)
            else:
                # Use default encoder from config
                encoder_name = self.config.get_yaml_config(
                    "memory_config.yaml", 
                    "tokens.tiktoken_model", 
                    "cl100k_base"
                )
                encoder = self._encoders.get(encoder_name)
            
            if not encoder:
                logger.warning(f"No encoder found for model {model}, using approximate counting")
                return self.count_tokens_approximate(text)
            
            # Count tokens
            tokens = encoder.encode(text)
            
            return TokenCount(
                tokens=len(tokens),
                characters=len(text),
                words=len(text.split()),
                method="tiktoken",
                model=model
            )
            
        except Exception as e:
            logger.error(f"tiktoken counting failed: {e}, falling back to approximate")
            return self.count_tokens_approximate(text)
    
    def count_tokens_approximate(self, text: str) -> TokenCount:
        """
        Count tokens using approximate character-based method.
        
        Args:
            text: Text to count tokens for
            
        Returns:
            Token count result
        """
        # Get ratio from config
        char_per_token = self.config.get_yaml_config(
            "memory_config.yaml",
            "tokens.approximate_ratio",
            4
        )
        
        characters = len(text)
        words = len(text.split())
        tokens = max(1, characters // char_per_token)
        
        return TokenCount(
            tokens=tokens,
            characters=characters,
            words=words,
            method="approximate"
        )
    
    def count_tokens_character_based(self, text: str) -> TokenCount:
        """
        Count tokens using simple character-based method.
        
        Args:
            text: Text to count tokens for
            
        Returns:
            Token count result
        """
        characters = len(text)
        words = len(text.split())
        tokens = characters  # 1:1 ratio for conservative estimation
        
        return TokenCount(
            tokens=tokens,
            characters=characters,
            words=words,
            method="character_based"
        )
    
    def count_tokens(self, text: str, model: Optional[str] = None, method: Optional[str] = None) -> TokenCount:
        """
        Count tokens using specified or configured method.
        
        Args:
            text: Text to count tokens for
            model: Model name for encoder selection
            method: Counting method ("tiktoken", "approximate", "character_based")
            
        Returns:
            Token count result
        """
        if not text:
            return TokenCount(tokens=0, characters=0, words=0, method="empty")
        
        # Determine method
        if not method:
            method = self.config.get_yaml_config(
                "memory_config.yaml",
                "tokens.counter",
                "tiktoken"
            )
        
        # Count tokens based on method
        if method == "tiktoken":
            return self.count_tokens_tiktoken(text, model)
        elif method == "approximate":
            return self.count_tokens_approximate(text)
        elif method == "character_based":
            return self.count_tokens_character_based(text)
        else:
            logger.warning(f"Unknown counting method {method}, using tiktoken")
            return self.count_tokens_tiktoken(text, model)
    
    def estimate_tokens_for_messages(self, messages: List[Dict[str, str]], model: Optional[str] = None) -> TokenCount:
        """
        Estimate tokens for a list of chat messages.
        
        Args:
            messages: List of message dictionaries with role and content
            model: Model name for accurate counting
            
        Returns:
            Total token count for all messages
        """
        total_tokens = 0
        total_chars = 0
        total_words = 0
        
        for message in messages:
            # Count content tokens
            content = message.get("content", "")
            role = message.get("role", "")
            
            content_count = self.count_tokens(content, model)
            role_count = self.count_tokens(role, model)
            
            # Add message formatting overhead (estimated)
            message_tokens = content_count.tokens + role_count.tokens + 4  # Overhead for message structure
            
            total_tokens += message_tokens
            total_chars += content_count.characters + role_count.characters
            total_words += content_count.words + role_count.words
        
        return TokenCount(
            tokens=total_tokens,
            characters=total_chars,
            words=total_words,
            method=content_count.method if messages else "empty",
            model=model
        )
    
    def truncate_text_to_tokens(self, text: str, max_tokens: int, model: Optional[str] = None, preserve_end: bool = False) -> str:
        """
        Truncate text to fit within token limit.
        
        Args:
            text: Text to truncate
            max_tokens: Maximum number of tokens
            model: Model name for accurate counting
            preserve_end: Whether to preserve the end of the text instead of beginning
            
        Returns:
            Truncated text
        """
        current_count = self.count_tokens(text, model)
        
        if current_count.tokens <= max_tokens:
            return text
        
        # Binary search for optimal truncation point
        if preserve_end:
            # Preserve end of text
            start = 0
            end = len(text)
            
            while start < end:
                mid = (start + end) // 2
                truncated = text[mid:]
                
                if self.count_tokens(truncated, model).tokens <= max_tokens:
                    end = mid
                else:
                    start = mid + 1
            
            return text[start:]
        else:
            # Preserve beginning of text
            start = 0
            end = len(text)
            
            while start < end:
                mid = (start + end + 1) // 2
                truncated = text[:mid]
                
                if self.count_tokens(truncated, model).tokens <= max_tokens:
                    start = mid
                else:
                    end = mid - 1
            
            return text[:start]
    
    def split_text_by_tokens(self, text: str, chunk_size: int, overlap: int = 0, model: Optional[str] = None) -> List[str]:
        """
        Split text into chunks of specified token size.
        
        Args:
            text: Text to split
            chunk_size: Maximum tokens per chunk
            overlap: Number of tokens to overlap between chunks
            model: Model name for accurate counting
            
        Returns:
            List of text chunks
        """
        if not text:
            return []
        
        # Simple sentence-based splitting with token counting
        sentences = re.split(r'[.!?]+', text)
        chunks = []
        current_chunk = ""
        
        for sentence in sentences:
            sentence = sentence.strip()
            if not sentence:
                continue
            
            # Check if adding this sentence exceeds limit
            test_chunk = current_chunk + " " + sentence if current_chunk else sentence
            token_count = self.count_tokens(test_chunk, model)
            
            if token_count.tokens <= chunk_size:
                current_chunk = test_chunk
            else:
                # Start new chunk
                if current_chunk:
                    chunks.append(current_chunk.strip())
                
                # Handle overlap
                if overlap > 0 and chunks:
                    overlap_text = self.truncate_text_to_tokens(
                        current_chunk, 
                        overlap, 
                        model, 
                        preserve_end=True
                    )
                    current_chunk = overlap_text + " " + sentence
                else:
                    current_chunk = sentence
        
        # Add final chunk
        if current_chunk:
            chunks.append(current_chunk.strip())
        
        return chunks
    
    def get_token_budget_info(self, memory_blocks: Dict[str, str], model: Optional[str] = None) -> Dict[str, Any]:
        """
        Get token budget information for memory blocks.
        
        Args:
            memory_blocks: Dictionary mapping block names to their content
            model: Model name for accurate counting
            
        Returns:
            Dictionary with token budget analysis
        """
        memory_config = self.config.get_memory_config()
        
        block_info = {}
        total_tokens = 0
        
        for block_name, content in memory_blocks.items():
            token_count = self.count_tokens(content, model)
            limit = memory_config.core_block_token_limit
            
            block_info[block_name] = {
                "current_tokens": token_count.tokens,
                "limit_tokens": limit,
                "utilization": round((token_count.tokens / limit) * 100, 1) if limit > 0 else 0,
                "characters": token_count.characters,
                "words": token_count.words,
                "exceeds_limit": token_count.tokens > limit,
                "tokens_over_limit": max(0, token_count.tokens - limit)
            }
            
            total_tokens += token_count.tokens
        
        return {
            "blocks": block_info,
            "total": {
                "current_tokens": total_tokens,
                "limit_tokens": memory_config.core_block_token_limit * len(memory_blocks),
                "utilization": round((total_tokens / (memory_config.core_block_token_limit * len(memory_blocks))) * 100, 1) if memory_blocks else 0,
                "blocks_over_limit": sum(1 for info in block_info.values() if info["exceeds_limit"])
            },
            "method": token_count.method if memory_blocks else "none",
            "model": model
        }


# Global token counter instance
token_counter = TokenCounter()


def count_tokens(text: str, model: Optional[str] = None, method: Optional[str] = None) -> TokenCount:
    """
    Convenience function to count tokens in text.
    
    Args:
        text: Text to count tokens for
        model: Model name for encoder selection
        method: Counting method override
        
    Returns:
        Token count result
    """
    return token_counter.count_tokens(text, model, method)


def estimate_message_tokens(messages: List[Dict[str, str]], model: Optional[str] = None) -> TokenCount:
    """
    Convenience function to estimate tokens for chat messages.
    
    Args:
        messages: List of message dictionaries
        model: Model name for accurate counting
        
    Returns:
        Total token count for messages
    """
    return token_counter.estimate_tokens_for_messages(messages, model)


def truncate_to_tokens(text: str, max_tokens: int, model: Optional[str] = None, preserve_end: bool = False) -> str:
    """
    Convenience function to truncate text to token limit.
    
    Args:
        text: Text to truncate
        max_tokens: Maximum tokens allowed
        model: Model name for accurate counting
        preserve_end: Whether to preserve end instead of beginning
        
    Returns:
        Truncated text
    """
    return token_counter.truncate_text_to_tokens(text, max_tokens, model, preserve_end)


def split_by_tokens(text: str, chunk_size: int, overlap: int = 0, model: Optional[str] = None) -> List[str]:
    """
    Convenience function to split text by token chunks.
    
    Args:
        text: Text to split
        chunk_size: Maximum tokens per chunk
        overlap: Overlap tokens between chunks
        model: Model name for accurate counting
        
    Returns:
        List of text chunks
    """
    return token_counter.split_text_by_tokens(text, chunk_size, overlap, model)


def analyze_memory_tokens(memory_blocks: Dict[str, str], model: Optional[str] = None) -> Dict[str, Any]:
    """
    Convenience function to analyze token usage in memory blocks.
    
    Args:
        memory_blocks: Dictionary of memory block content
        model: Model name for accurate counting
        
    Returns:
        Token budget analysis
    """
    return token_counter.get_token_budget_info(memory_blocks, model)