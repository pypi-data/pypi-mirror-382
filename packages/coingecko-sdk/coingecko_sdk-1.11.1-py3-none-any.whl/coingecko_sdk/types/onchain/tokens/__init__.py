# File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.

from __future__ import annotations

from .info_recently_updated_get_params import InfoRecentlyUpdatedGetParams as InfoRecentlyUpdatedGetParams
from .info_recently_updated_get_response import InfoRecentlyUpdatedGetResponse as InfoRecentlyUpdatedGetResponse
