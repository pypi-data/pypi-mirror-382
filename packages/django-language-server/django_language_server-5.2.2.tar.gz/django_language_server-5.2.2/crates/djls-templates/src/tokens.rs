use djls_source::Span;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum TagDelimiter {
    Block,
    Variable,
    Comment,
}

impl TagDelimiter {
    pub const CHAR_OPEN: char = '{';
    pub const LENGTH: usize = 2;
    pub const LENGTH_U32: u32 = 2;

    #[must_use]
    pub fn from_input(input: &str) -> Option<Self> {
        let bytes = input.as_bytes();

        if bytes.len() < Self::LENGTH {
            return None;
        }

        if bytes[0] != Self::CHAR_OPEN as u8 {
            return None;
        }

        match bytes[1] {
            b'%' => Some(Self::Block),
            b'{' => Some(Self::Variable),
            b'#' => Some(Self::Comment),
            _ => None,
        }
    }

    #[must_use]
    pub fn opener(self) -> &'static str {
        match self {
            Self::Block => "{%",
            Self::Variable => "{{",
            Self::Comment => "{#",
        }
    }

    #[must_use]
    pub fn closer(self) -> &'static str {
        match self {
            Self::Block => "%}",
            Self::Variable => "}}",
            Self::Comment => "#}",
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Token {
    Block { content: String, span: Span },
    Comment { content: String, span: Span },
    Error { content: String, span: Span },
    Eof,
    Newline { span: Span },
    Text { content: String, span: Span },
    Variable { content: String, span: Span },
    Whitespace { span: Span },
}

impl Token {
    /// Get the content text for content-bearing tokens
    #[must_use]
    pub fn content(&self) -> String {
        match self {
            Token::Block { content, .. }
            | Token::Comment { content, .. }
            | Token::Error { content, .. }
            | Token::Text { content, .. }
            | Token::Variable { content, .. } => content.clone(),
            Token::Whitespace { span, .. } => " ".repeat(span.length_usize()),
            Token::Newline { span, .. } => {
                if span.length() == 2 {
                    "\r\n".to_string()
                } else {
                    "\n".to_string()
                }
            }
            Token::Eof => String::new(),
        }
    }

    /// Get the lexeme as it appears in source
    #[must_use]
    pub fn lexeme(&self) -> String {
        match self {
            Token::Block { content, .. } => format!(
                "{} {} {}",
                TagDelimiter::Block.opener(),
                content,
                TagDelimiter::Block.closer()
            ),
            Token::Variable { content, .. } => format!(
                "{} {} {}",
                TagDelimiter::Variable.opener(),
                content,
                TagDelimiter::Variable.closer()
            ),
            Token::Comment { content, .. } => format!(
                "{} {} {}",
                TagDelimiter::Comment.opener(),
                content,
                TagDelimiter::Comment.closer()
            ),
            Token::Text { content, .. } | Token::Error { content, .. } => content.clone(),
            Token::Whitespace { span, .. } => " ".repeat(span.length_usize()),
            Token::Newline { span, .. } => {
                if span.length() == 2 {
                    "\r\n".to_string()
                } else {
                    "\n".to_string()
                }
            }
            Token::Eof => String::new(),
        }
    }

    #[must_use]
    pub fn offset(&self) -> Option<u32> {
        match self {
            Token::Block { span, .. }
            | Token::Comment { span, .. }
            | Token::Error { span, .. }
            | Token::Variable { span, .. } => {
                Some(span.start().saturating_sub(TagDelimiter::LENGTH_U32))
            }
            Token::Text { span, .. }
            | Token::Whitespace { span, .. }
            | Token::Newline { span, .. } => Some(span.start()),
            Token::Eof => None,
        }
    }

    /// Get the length of the token content
    #[must_use]
    pub fn length(&self) -> u32 {
        let len = match self {
            Token::Block { content, .. }
            | Token::Comment { content, .. }
            | Token::Error { content, .. }
            | Token::Text { content, .. }
            | Token::Variable { content, .. } => content.len(),
            Token::Whitespace { span, .. } | Token::Newline { span, .. } => span.length_usize(),
            Token::Eof => 0,
        };
        u32::try_from(len).unwrap_or(u32::MAX)
    }

    #[must_use]
    pub fn full_span(&self) -> Option<Span> {
        match self {
            Token::Block { span, .. }
            | Token::Comment { span, .. }
            | Token::Variable { span, .. } => {
                Some(span.expand(TagDelimiter::LENGTH_U32, TagDelimiter::LENGTH_U32))
            }
            Token::Error { span, .. } => Some(span.expand(TagDelimiter::LENGTH_U32, 0)),
            Token::Newline { span, .. }
            | Token::Text { span, .. }
            | Token::Whitespace { span, .. } => Some(*span),
            Token::Eof => None,
        }
    }

    #[must_use]
    pub fn content_span(&self) -> Option<Span> {
        match self {
            Token::Block { span, .. }
            | Token::Comment { span, .. }
            | Token::Error { span, .. }
            | Token::Text { span, .. }
            | Token::Variable { span, .. }
            | Token::Whitespace { span, .. }
            | Token::Newline { span, .. } => Some(*span),
            Token::Eof => None,
        }
    }

    #[must_use]
    pub fn full_span_or_fallback(&self) -> Span {
        self.full_span()
            .unwrap_or_else(|| self.content_span_or_fallback())
    }

    #[must_use]
    pub fn content_span_or_fallback(&self) -> Span {
        self.content_span()
            .unwrap_or_else(|| Span::new(self.offset().unwrap_or(0), self.length()))
    }

    #[must_use]
    pub fn spans(&self) -> (Span, Span) {
        let content = self.content_span_or_fallback();
        let full = self.full_span().unwrap_or(content);
        (content, full)
    }
}

#[cfg(test)]
#[derive(Debug, serde::Serialize)]
pub enum TokenSnapshot {
    Block {
        content: String,
        span: (u32, u32),
        full_span: (u32, u32),
    },
    Comment {
        content: String,
        span: (u32, u32),
        full_span: (u32, u32),
    },
    Eof,
    Error {
        content: String,
        span: (u32, u32),
        full_span: (u32, u32),
    },
    Newline {
        span: (u32, u32),
    },
    Text {
        content: String,
        span: (u32, u32),
        full_span: (u32, u32),
    },
    Variable {
        content: String,
        span: (u32, u32),
        full_span: (u32, u32),
    },
    Whitespace {
        span: (u32, u32),
    },
}

#[cfg(test)]
impl Token {
    /// ## Panics
    ///
    /// This may panic on the `full_span` calls, but it's only used in testing,
    /// so it's all good.
    #[must_use]
    pub fn to_snapshot(&self) -> TokenSnapshot {
        match self {
            Token::Block { span, .. } => TokenSnapshot::Block {
                content: self.content(),
                span: span.into(),
                full_span: self.full_span().unwrap().into(),
            },
            Token::Comment { span, .. } => TokenSnapshot::Comment {
                content: self.content(),
                span: span.into(),
                full_span: self.full_span().unwrap().into(),
            },
            Token::Eof => TokenSnapshot::Eof,
            Token::Error { span, .. } => TokenSnapshot::Error {
                content: self.content(),
                span: span.into(),
                full_span: self.full_span().unwrap().into(),
            },
            Token::Newline { span } => TokenSnapshot::Newline { span: span.into() },
            Token::Text { span, .. } => TokenSnapshot::Text {
                content: self.content(),
                span: span.into(),
                full_span: span.into(),
            },
            Token::Variable { span, .. } => TokenSnapshot::Variable {
                content: self.content(),
                span: span.into(),
                full_span: self.full_span().unwrap().into(),
            },
            Token::Whitespace { span } => TokenSnapshot::Whitespace { span: span.into() },
        }
    }
}

#[cfg(test)]
pub struct TokenSnapshotVec(pub Vec<Token>);

#[cfg(test)]
impl TokenSnapshotVec {
    #[must_use]
    pub fn to_snapshot(&self) -> Vec<TokenSnapshot> {
        self.0.iter().map(Token::to_snapshot).collect()
    }
}

#[derive(Debug, Clone)]
pub struct TokenStream(Vec<Token>);

impl TokenStream {
    const CHARS_PER_TOKEN: usize = 6;
    const MIN_CAPACITY: usize = 32;
    const MAX_CAPACITY: usize = 1024;

    #[must_use]
    pub fn with_estimated_capacity(source: &str) -> Self {
        let capacity =
            (source.len() / Self::CHARS_PER_TOKEN).clamp(Self::MIN_CAPACITY, Self::MAX_CAPACITY);
        Self(Vec::with_capacity(capacity))
    }

    #[inline]
    pub fn push(&mut self, token: Token) {
        self.0.push(token);
    }

    /// Get the number of tokens in the stream.
    #[must_use]
    pub fn len(&self) -> usize {
        self.0.len()
    }

    /// Get the number of content tokens (excluding EOF).
    #[must_use]
    pub fn content_len(&self) -> usize {
        self.0.len().saturating_sub(1)
    }

    /// Check if stream is empty.
    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.0.is_empty()
    }
}

impl From<TokenStream> for Vec<Token> {
    fn from(val: TokenStream) -> Self {
        val.0
    }
}

impl IntoIterator for TokenStream {
    type Item = Token;
    type IntoIter = std::vec::IntoIter<Token>;

    fn into_iter(self) -> Self::IntoIter {
        self.0.into_iter()
    }
}
