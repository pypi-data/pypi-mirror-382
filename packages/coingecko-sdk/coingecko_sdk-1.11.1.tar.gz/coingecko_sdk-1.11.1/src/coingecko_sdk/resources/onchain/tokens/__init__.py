# File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.

from .tokens import (
    TokensResource,
    AsyncTokensResource,
    TokensResourceWithRawResponse,
    AsyncTokensResourceWithRawResponse,
    TokensResourceWithStreamingResponse,
    AsyncTokensResourceWithStreamingResponse,
)
from .info_recently_updated import (
    InfoRecentlyUpdatedResource,
    AsyncInfoRecentlyUpdatedResource,
    InfoRecentlyUpdatedResourceWithRawResponse,
    AsyncInfoRecentlyUpdatedResourceWithRawResponse,
    InfoRecentlyUpdatedResourceWithStreamingResponse,
    AsyncInfoRecentlyUpdatedResourceWithStreamingResponse,
)

__all__ = [
    "InfoRecentlyUpdatedResource",
    "AsyncInfoRecentlyUpdatedResource",
    "InfoRecentlyUpdatedResourceWithRawResponse",
    "AsyncInfoRecentlyUpdatedResourceWithRawResponse",
    "InfoRecentlyUpdatedResourceWithStreamingResponse",
    "AsyncInfoRecentlyUpdatedResourceWithStreamingResponse",
    "TokensResource",
    "AsyncTokensResource",
    "TokensResourceWithRawResponse",
    "AsyncTokensResourceWithRawResponse",
    "TokensResourceWithStreamingResponse",
    "AsyncTokensResourceWithStreamingResponse",
]
