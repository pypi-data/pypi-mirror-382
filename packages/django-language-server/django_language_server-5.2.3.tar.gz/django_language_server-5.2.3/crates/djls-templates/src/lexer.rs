use djls_source::Span;
use memchr::memchr3;
use memchr::memmem;

use crate::tokens::TagDelimiter;
use crate::tokens::Token;
use crate::tokens::TokenStream;

pub struct Lexer {
    source: String,
    start: usize,
    current: usize,
}

impl Lexer {
    #[must_use]
    pub fn new(source: &str) -> Self {
        Lexer {
            source: String::from(source),
            start: 0,
            current: 0,
        }
    }

    pub fn tokenize(&mut self) -> Vec<Token> {
        let mut tokens = TokenStream::with_estimated_capacity(&self.source);

        while !self.is_at_end() {
            self.start = self.current;

            let token =
                match self.peek() {
                    TagDelimiter::CHAR_OPEN => {
                        let remaining = self.remaining_source();

                        match TagDelimiter::from_input(remaining) {
                            Some(TagDelimiter::Block) => self
                                .lex_django_tag(TagDelimiter::Block, |content, span| {
                                    Token::Block { content, span }
                                }),
                            Some(TagDelimiter::Variable) => self
                                .lex_django_tag(TagDelimiter::Variable, |content, span| {
                                    Token::Variable { content, span }
                                }),
                            Some(TagDelimiter::Comment) => self
                                .lex_django_tag(TagDelimiter::Comment, |content, span| {
                                    Token::Comment { content, span }
                                }),
                            None => self.lex_text(),
                        }
                    }
                    c if c.is_whitespace() => self.lex_whitespace(c),
                    _ => self.lex_text(),
                };

            tokens.push(token);
        }

        tokens.push(Token::Eof);

        tokens.into()
    }

    fn lex_django_tag(
        &mut self,
        delimiter: TagDelimiter,
        token_fn: impl FnOnce(String, Span) -> Token,
    ) -> Token {
        let content_start = self.start + TagDelimiter::LENGTH;

        self.consume_n(TagDelimiter::LENGTH);

        match self.consume_until_delimiter(delimiter.closer()) {
            Ok(text) => {
                let len = text.len();
                let span = Span::saturating_from_parts_usize(content_start, len);
                self.consume_n(delimiter.closer().len());
                token_fn(text, span)
            }
            Err(err_text) => {
                let len = err_text.len();
                let span = if len == 0 {
                    Span::saturating_from_bounds_usize(content_start, self.current)
                } else {
                    Span::saturating_from_parts_usize(content_start, len)
                };
                Token::Error {
                    content: err_text,
                    span,
                }
            }
        }
    }

    fn lex_whitespace(&mut self, c: char) -> Token {
        self.consume();

        if c == '\n' || c == '\r' {
            if c == '\r' && self.peek() == '\n' {
                self.consume();
            }
            let span = Span::saturating_from_bounds_usize(self.start, self.current);
            return Token::Newline { span };
        }

        while !self.is_at_end() {
            let remaining = self.remaining_source().as_bytes();

            match remaining.first() {
                Some(&b'\n' | &b'\r') | None => break,
                Some(&b' ' | &b'\t') => self.current += 1,
                Some(_) => {
                    if !self.peek().is_whitespace() {
                        break;
                    }
                    self.consume();
                }
            }
        }

        let span = Span::saturating_from_bounds_usize(self.start, self.current);
        Token::Whitespace { span }
    }

    fn lex_text(&mut self) -> Token {
        let text_start = self.current;
        self.current += self.consume_until_stop_char();
        let text = self.consumed_source_from(text_start);
        let span = Span::saturating_from_bounds_usize(self.start, self.current);
        Token::Text {
            content: text.to_string(),
            span,
        }
    }

    #[inline]
    fn peek(&self) -> char {
        self.remaining_source().chars().next().unwrap_or('\0')
    }

    #[inline]
    fn remaining_source(&self) -> &str {
        &self.source[self.current..]
    }

    #[inline]
    fn consumed_source_from(&self, start: usize) -> &str {
        &self.source[start..self.current]
    }

    #[inline]
    fn is_at_end(&self) -> bool {
        self.current >= self.source.len()
    }

    #[inline]
    fn consume(&mut self) {
        if let Some(ch) = self.remaining_source().chars().next() {
            self.current += ch.len_utf8();
        }
    }

    fn consume_n(&mut self, count: usize) {
        for _ in 0..count {
            self.consume();
        }
    }

    fn consume_until_delimiter(&mut self, delimiter: &str) -> Result<String, String> {
        let offset = self.current;

        if let Some(pos) = memmem::find(self.remaining_source().as_bytes(), delimiter.as_bytes()) {
            self.current += pos;
            return Ok(self.consumed_source_from(offset).to_string());
        }

        self.current += self.consume_until_stop_char();
        Err(self.consumed_source_from(offset).to_string())
    }

    fn consume_until_stop_char(&self) -> usize {
        let mut offset = 0;
        let max = self.source.len() - self.current;

        while offset < max {
            let remaining = &self.remaining_source()[offset..];

            match memchr3(b'{', b'\n', b'\r', remaining.as_bytes()) {
                None => {
                    offset = max;
                    break;
                }
                Some(pos) => {
                    let is_newline = matches!(remaining.as_bytes()[pos], b'\n' | b'\r');
                    let is_django_delimiter = TagDelimiter::from_input(&remaining[pos..]).is_some();

                    if is_newline || is_django_delimiter {
                        offset += pos;
                        break;
                    }

                    offset += pos + 1;
                }
            }
        }

        offset
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tokens::TokenSnapshotVec;

    #[test]
    fn test_tokenize_html() {
        let source = r#"<div class="container" id="main" disabled></div>"#;
        let mut lexer = Lexer::new(source);
        let tokens = lexer.tokenize();
        let snapshot = TokenSnapshotVec(tokens).to_snapshot();
        insta::assert_yaml_snapshot!(snapshot);
    }

    #[test]
    fn test_tokenize_django_variable() {
        let source = "{{ user.name|default:\"Anonymous\"|title }}";
        let mut lexer = Lexer::new(source);
        let tokens = lexer.tokenize();
        let snapshot = TokenSnapshotVec(tokens).to_snapshot();
        insta::assert_yaml_snapshot!(snapshot);
    }

    #[test]
    fn test_tokenize_django_block() {
        let source = "{% if user.is_staff %}Admin{% else %}User{% endif %}";
        let mut lexer = Lexer::new(source);
        let tokens = lexer.tokenize();
        let snapshot = TokenSnapshotVec(tokens).to_snapshot();
        insta::assert_yaml_snapshot!(snapshot);
    }

    #[test]
    fn test_tokenize_comments() {
        let source = r"<!-- HTML comment -->
{# Django comment #}
<script>
    // JS single line comment
    /* JS multi-line
       comment */
</script>
<style>
    /* CSS comment */
</style>";
        let mut lexer = Lexer::new(source);
        let tokens = lexer.tokenize();
        let snapshot = TokenSnapshotVec(tokens).to_snapshot();
        insta::assert_yaml_snapshot!(snapshot);
    }

    #[test]
    fn test_tokenize_script() {
        let source = r#"<script type="text/javascript">
    // Single line comment
    const x = 1;
    /* Multi-line
       comment */
    console.log(x);
</script>"#;
        let mut lexer = Lexer::new(source);
        let tokens = lexer.tokenize();
        let snapshot = TokenSnapshotVec(tokens).to_snapshot();
        insta::assert_yaml_snapshot!(snapshot);
    }

    #[test]
    fn test_tokenize_style() {
        let source = r#"<style type="text/css">
    /* Header styles */
    .header {
        color: blue;
    }
</style>"#;
        let mut lexer = Lexer::new(source);
        let tokens = lexer.tokenize();
        let snapshot = TokenSnapshotVec(tokens).to_snapshot();
        insta::assert_yaml_snapshot!(snapshot);
    }

    #[test]
    fn test_tokenize_nested_delimiters() {
        let source = r"{{ user.name }}
{% if true %}
{# comment #}
<!-- html comment -->
<div>text</div>";
        let mut lexer = Lexer::new(source);
        let tokens = lexer.tokenize();
        let snapshot = TokenSnapshotVec(tokens).to_snapshot();
        insta::assert_yaml_snapshot!(snapshot);
    }

    #[test]
    fn test_tokenize_everything() {
        let source = r#"<!DOCTYPE html>
<html>
<head>
    <style type="text/css">
        /* Style header */
        .header { color: blue; }
    </style>
    <script type="text/javascript">
        // Init app
        const app = {
            /* Config */
            debug: true
        };
    </script>
</head>
<body>
    <!-- Header section -->
    <div class="header" id="main" data-value="123" disabled>
        {% if user.is_authenticated %}
            {# Welcome message #}
            <h1>Welcome, {{ user.name|default:"Guest"|title }}!</h1>
            {% if user.is_staff %}
                <span>Admin</span>
            {% else %}
                <span>User</span>
            {% endif %}
        {% endif %}
    </div>
</body>
</html>"#;
        let mut lexer = Lexer::new(source);
        let tokens = lexer.tokenize();
        let snapshot = TokenSnapshotVec(tokens).to_snapshot();
        insta::assert_yaml_snapshot!(snapshot);
    }

    #[test]
    fn test_tokenize_unclosed_style() {
        let source = "<style>body { color: blue; ";
        let mut lexer = Lexer::new(source);
        let tokens = lexer.tokenize();
        let snapshot = TokenSnapshotVec(tokens).to_snapshot();
        insta::assert_yaml_snapshot!(snapshot);
    }
}
