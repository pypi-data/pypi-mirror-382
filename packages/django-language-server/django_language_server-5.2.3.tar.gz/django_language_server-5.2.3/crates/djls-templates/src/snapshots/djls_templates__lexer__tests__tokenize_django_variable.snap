---
source: crates/djls-templates/src/lexer.rs
expression: snapshot
---
- Variable:
    content: " user.name|default:\"Anonymous\"|title "
    span:
      - 2
      - 37
    full_span:
      - 0
      - 41
- Eof
